### Project Overview
- You are assigned to retrieve 5 years of data on secirities belinging to the S&P500, Nasdaq, and Russell
- You will Use the ADJ Close and Volume(daily change)  of these companie to try to predict the value of the UVXY 
- Use feature engineering to  try to generate more features for the data set, how do these features coorelate to UVXY
- Does these valuse have any predictive abilities on UVXY?
- Keep in mind you will be working with alot of companies and features so consider a method to reduce these features, may be attempt a PCA and see how the results will pan out

## Importing Libraries

In [2]:
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt
import seaborn as sns
from pandas_datareader import data as web

## importing Tickers 
- WE will import all tickers from Nasdaq, S&P, and Russell as of 2020

In [12]:
sandp = pd.read_csv("resources/companies.csv")

In [13]:
len(sandp["Symbol"])

505

In [22]:
russell = pd.read_excel("resources/Russle_2000_2020.xlsx")

In [23]:
russell_tickers = list(russell["Ticker"])

In [25]:
len(russell_tickers)

1999

In [26]:
russell_zip = dict(zip(russell["Ticker"], russell["Name"]))

In [30]:
nasdaq = pd.read_excel("resources/nasdaq_100_stocks.xlsx")

In [31]:
nasdaq_tickers = nasdaq["Ticker"]

In [33]:
len(nasdaq_tickers)

103

In [34]:
nasdaq_zipp = dict(zip(nasdaq["Ticker"], nasdaq["Name"]))

## Zipping
- We will zip the ticker and the company's name together if we need to query the company's name at some point

In [17]:
sandpZip = dict(zip(sandp["Symbol"], sandp["Name"]))

In [39]:
sandp_tickers.extend(nasdaq_tickers)

In [40]:
len(sandp_tickers)

608

In [47]:
sandp_tickers.extend(russell_tickers)

In [49]:
len(sandp_tickers)

4606

In [50]:
tickers = sandp_tickers

## Getting Stock Data for Each ticker 
- We will begin with the adj Close

In [53]:
df = pd.DataFrame()

### Saving
- because this is a taxing task to gather a little over 4000 companies Adj Close over a 5 year period 
- once we have done it we will save the data frame to a csv file for future reference
- WE do not want to have to fetchingthe data every time
- We will then import the data below

In [54]:
##failed = []
##passed= []
##for x in tickers:
   ## try:
       ## df[x] = web.DataReader(name=x, data_source="yahoo", start = "2016-1-1")["Adj Close"]
       ## passed.append(x)
    ##except(IOError, KeyError):
        ##failed.append(x)

In [55]:
len(passed)

4235

In [57]:
##df.to_csv("resources/securityData.csv")

In [60]:
df = pd.read_csv("resources/securityData.csv", index_col="Date")

### Evaluating the DF

In [61]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1288 entries, 2016-01-04 to 2021-02-12
Columns: 2321 entries, MMM to ZYXI
dtypes: float64(2321)
memory usage: 22.8+ MB


### Overview
- We are looking at approximately 2321 companies to date
- 1288 trading days

In [68]:
len(df.columns)

2321

### Number of years

In [67]:
len(df) / 252

5.111111111111111

## Lets analyze the data 
- Will check for missing data 
- we will also do some visualization
- we will drop companies that has too many missing values

In [69]:
df.describe()

MMM          AOS  ...         ZYNE         ZYXI
count  1288.000000  1288.000000  ...  1288.000000  1288.000000
mean    169.359206    48.308186  ...     9.213152     5.829549
std      21.905798     7.507483  ...     4.904679     6.256987
min     113.730507    28.644878  ...     2.890000     0.155774
25%     153.929790    43.635086  ...     5.337500     0.350491
50%     166.367989    47.405514  ...     8.150000     3.115472
75%     184.934326    54.572262  ...    11.560000     9.540000
max     233.063889    64.031891  ...    25.190001    28.580000

[8 rows x 2321 columns]

## Chceking for missing data 
- appears we have companies that are missing a good amount of data
- We will need to decide what we will do with these companies as we move through our analysis 
- IF there are too many values missing we will need to replace them or drop them fro mthe list

In [77]:
df.isnull().sum().sort_values(ascending = False).head(20)

XOG     1271
AVDR    1257
PX      1251
AI      1243
AAN     1234
CRC     1214
NWE     1072
ARNC    1068
PRMW    1048
PNTG     942
XL       922
CFB      910
KRUS     900
NOVA     895
PROS     895
CSTL     895
HCAT     895
AMK      890
PHR      890
MIRM     890
dtype: int64

### Importing data for UVXY

In [79]:
uvxy = web.DataReader(name="UVXY", data_source="yahoo", start = "2016-1-1")["Adj Close"]

In [101]:
df["UVXY"] = uvxy

In [122]:
df

MMM        AOS        ABT  ...  ZYNE      ZYXI   UVXY
Date                                                   ...                       
2016-01-04           125.897713  34.700645  38.764435  ...  9.67  0.360226    NaN
2016-01-05           126.446503  34.506790  38.755398  ...  9.25  0.399170    NaN
2016-01-06           123.899742  33.823669  38.430336  ...  9.10  0.428377    NaN
2016-01-07           120.881340  32.171246  37.509308  ...  8.75  0.399170    NaN
2016-01-08           120.469727  31.026575  36.723721  ...  8.77  0.369962    NaN
...                         ...        ...        ...  ...   ...       ...    ...
2021-02-08 00:00:00         NaN        NaN        NaN  ...   NaN       NaN   9.98
2021-02-09 00:00:00         NaN        NaN        NaN  ...   NaN       NaN  10.01
2021-02-10 00:00:00         NaN        NaN        NaN  ...   NaN       NaN  10.25
2021-02-11 00:00:00         NaN        NaN        NaN  ...   NaN       NaN   9.91
2021-02-12 00:00:00         NaN        NaN        NaN  ...   NaN       NaN   9.42

[2576 rows x 2322 columns]

In [125]:
passed.append("UVXY")

In [126]:
data = pd.DataFrame()

In [129]:
failed = []
passed= []
for x in df.columns:
    try:
        data[x] = web.DataReader(name=x, data_source="yahoo", start = "2019-1-1")["Adj Close"]
        passed.append(x)
    except(IOError, KeyError):
        failed.append(x)

In [130]:
data

MMM        AOS         ABT  ...   ZYNE       ZYXI       UVXY
Date                                           ...                             
2019-01-02  176.638840  41.601696   67.034943  ...  3.720   2.890000  77.919998
2019-01-03  169.987732  40.596821   63.871284  ...  3.670   2.820000  83.300003
2019-01-04  176.981140  42.223755   65.694260  ...  4.000   2.950000  73.540001
2019-01-07  176.574112  42.979805   66.678070  ...  4.000   2.960000  70.949997
2019-01-08  177.314148  44.032524   65.877518  ...  4.120   2.940000  68.800003
...                ...        ...         ...  ...    ...        ...        ...
2021-02-08  179.282608  61.380001  124.500000  ...  5.740  19.990000   9.980000
2021-02-09  179.461151  61.320000  125.150002  ...  5.955  19.990000  10.010000
2021-02-10  179.600006  61.220001  125.290001  ...  8.760  19.510000  10.250000
2021-02-11  177.210007  61.090000  126.839996  ...  6.110  19.280001   9.910000
2021-02-12  178.699997  60.369999  128.229996  ...  5.635  19.450001   9.420000

[534 rows x 2314 columns]

In [131]:
data.to_csv("resources/2019_currentData.csv")